### PR TITLE
IPAM pools followups

### PR DESCRIPTION
**Pull request by tklauser**
Address out-of-band review comments made on #22762. See commits for details.

For #25470

### CODE REVIEW

File: pkg/ipam/multipool.go [hunk 1]
```diff
@@ -453,20 +453,19 @@ func (m *multiPoolManager) allocateIP(ip net.IP, owner string, poolName Pool, fa
...
 
-func (m *multiPoolManager) releaseIP(ip net.IP, poolName Pool, family Family, upstreamSync bool) error {
+func (m *multiPoolManager) releaseIP(ip net.IP, poolName Pool, family Family, upstreamSync bool) {
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
 
 	pool := m.poolByFamilyLocked(poolName.String(), family)
 	if pool == nil {
-		return fmt.Errorf("unable to release IP %s of unknown pool %q (family %s)", ip, poolName, family)
+		return
```
Comment (gandro) [inline pkg/ipam/multipool.go]: I think we should at least emit a warning here. This is something that indicates desynced state after all. How could we have released a pool if an owner still thinks it has an IP allocated from it?

**Review by gandro — COMMENTED** [review]
Thanks!

File: pkg/ipam/multipool.go (hunk 1 above)
Comment (tklauser) [reply pkg/ipam/multipool.go]: Good point, added a warning.

File: pkg/ipam/service/allocator/interfaces.go [hunk 2]
```diff
@@ -11,11 +11,7 @@ type Interface interface {
 	AllocateNext() (int, bool, error)
 	Release(int) error
 	ForEach(func(int))
-
-	// For testing
```
Comment (aojea) [inline pkg/ipam/service/allocator/interfaces.go:15]: can you send the same change in kubernetes/kubernetes?

**Comment by aojea** [comment]
I'm not sure if the last commit is correct, let me dump my thoughts because I have the feeling that is conflating two allocators, the ipam used for cidrsset and the bitmap used for clusterips.

For the latter, since Kubernetes ClusterIP allocation happens on the apiserver synchronously, we can not fail so the errors are swallowed, but we have a repair loop that checks for leakages.
For the former, for the ipam, I think that error handling should be kept, but this is thinking about how the cidr IPAM works in Kubernetes, that assign the cidr to the pod.Spec.PodCIDR field.
This field is inmutable, that means if you try to release and you are not able to, there is a risk of having two nodes with the same PodCIDRs assigned, and the other operation to solve the problem is to delete the node object (with all the implications it have)

File: pkg/ipam/service/allocator/interfaces.go (hunk 2 above)
Comment (tklauser) [reply pkg/ipam/service/allocator/interfaces.go:15]: Opened https://github.com/kubernetes/kubernetes/pull/118171

**Comment by tklauser** [comment]
> I'm not sure if the last commit is correct, let me dump my thoughts because I have the feeling that is conflating two allocators, the cidrsset used for node ipam and the bitmap used for clusterips.
> 
> For the latter, since Kubernetes ClusterIP allocation happens on the apiserver synchronously, we can not fail (there is a long history here) so the errors are swallowed, but we have a repair loop that checks for leakages. For the former, for the ipam, I think that error handling should be kept, but this is thinking about how the cidr IPAM works in Kubernetes, that assign the cidr to the pod.Spec.PodCIDR field. This field is inmutable, that means if you try to release and you are not able to, there is a risk of having two nodes with the same PodCIDRs assigned, and the other operation to solve the problem is to delete the node object (with all the implications it have)

Thanks for pointing that out. I agree that the change is conflating the two allocators. I've modified the commit to only remove the error return from the `allocator.Interface` interface where the error already was always `nil`.

**Comment by tklauser** [comment]
/test-1.26-net-next

**Review by squeed — APPROVED** [review]

**Comment by tklauser** [comment]
/test-1.26-net-next

**Review by gandro — APPROVED** [review]

**Comment by tklauser** [comment]
/test-1.26-net-next

**Review by jibi — APPROVED** [review]

**Pull request merged**